File: android/src/main/java/org/ar/rtc/base/StreamingKit.kt
package org.ar.rtc.base

import android.content.Context
import org.ar.rtc.RtcEngine
import org.ar.rtc.rtmp.StreamingKit
import org.ar.rtc.rtmp.internal.PushMode
import org.ar.rtc.rtmp.jni.ISteamingEventHandler

class IStreamingKit{
    interface StreamingKitInterface{
        fun createInstance(params: Map<String, *>,callback: Callback)

        fun release(callback: Callback)

        fun pushStream(params: Map<String, *>, callback: Callback)

        fun unPushStream(callback: Callback)

        fun setMode(params: Map<String, *>, callback: Callback)

        fun setLiveTranscoding(params: Map<String, *>, callback: Callback)
    }
}


class StreamingKitManager:IStreamingKit.StreamingKitInterface{

    var streamingKit: StreamingKit? =null

    override fun createInstance(params: Map<String, *>,callback: Callback) {
        callback.resolve(params["engine"] as RtcEngine) { e ->
            streamingKit = StreamingKit.createInstance()
            streamingKit?.setRtcEngine(e)
            Unit
        }
    }

    override fun release(callback: Callback) {
        streamingKit?.release()
        callback.code(0)
    }

    override fun pushStream(params: Map<String, *>, callback: Callback) {
        callback.code(streamingKit?.pushStream(params["url"] as String))

    }

    override fun unPushStream(callback: Callback) {
        callback.code(streamingKit?.unPushStream())
    }

    override fun setMode(params: Map<String, *>, callback: Callback) {
        var mode = (params["mode"] as Number).toInt();
        callback.code(streamingKit?.setMode(PushMode.values()[mode]))
    }

    override fun setLiveTranscoding(params: Map<String, *>, callback: Callback) {
        callback.code(streamingKit?.setLiveTranscoding(mapToLiveTranscoding(params["transcoding"] as Map<*, *>)))
    }

}

File: ios/Classes/Base/StreamingKit.swift
//
//  StreamingKit.swift
//  ar_rtc_engine
//
//  Created by 余生丶 on 2021/1/7.
//

import Foundation
import ARtcKit

protocol StreamingKitInterface {
    
    func createInstance(_ params: NSDictionary, _ callback: Callback)

    func release(callback: Callback)

    func pushStream(_ params: NSDictionary, _ callback: Callback)

    func unPushStream(_ callback: Callback)

    func setMode(_ params: NSDictionary, _ callback: Callback)

    func setLiveTranscoding(params: NSDictionary, _ callback: Callback)
}

@objc
class StreamingKitManager: NSObject, StreamingKitInterface {
    
    private var emitter: (_ methodName: String, _ data: Dictionary<String, Any?>?) -> Void
    private(set) var streamingKit: ARStreamingKit?

    init(_ emitter: @escaping (_ methodName: String, _ data: Dictionary<String, Any?>?) -> Void) {
        self.emitter = emitter
    }

    func Release() {
        streamingKit = nil
    }
    
    @objc func createInstance(_ params: NSDictionary, _ callback: Callback) {
        callback.resolve(params["engine"] as? ARtcEngineKit) { (e) -> Any? in
            streamingKit = ARStreamingKit.init()
            streamingKit?.setRtcEngine(e)
            return 0
        }
    }
    
    @objc func release(callback: Callback) {
        streamingKit?.destroy()
        callback.code(0)
    }
    
    @objc func pushStream(_ params: NSDictionary, _ callback: Callback) {
        callback.code(streamingKit?.pushStream(params["url"] as! String))
    }
    
    @objc func unPushStream(_ callback: Callback) {
        callback.code(streamingKit?.unPushStream())
    }
    
    @objc func setMode(_ params: NSDictionary, _ callback: Callback) {
        callback.code(0)
    }
    
    @objc func setLiveTranscoding(params: NSDictionary, _ callback: Callback) {
        callback.code(streamingKit?.setLiveTranscoding(mapToLiveTranscoding(params["transcoding"] as! Dictionary)))
    }
}
